test/jdk/java/net/httpclient/UserAuthWithAuthenticator.java

```diff
@@ -0,0 +1,514 @@
+    }
+
+    // if useHeader is true, we expect the Authenticator was not called
+    // and the user set header used. If false, Authenticator must
+    // be called and the user set header not used.
+
+    static void h2Test(final boolean useHeader) throws Exception {
+        SSLContext ctx;
+        HttpTestServer h2s = null;
+        HttpClient client = null;
+        ExecutorService ex=null;
+        try {
+            ctx = new SimpleSSLContext().get();
+            ex = Executors.newCachedThreadPool();
+            InetAddress addr = InetAddress.getLoopbackAddress();
+
+            h2s = HttpTestServer.of(new Http2TestServer(addr, "::1", true, 0, ex,
+                    10, null, ctx, false));
+            AuthTestHandler h = new AuthTestHandler();
+            var context = h2s.addHandler(h, "/test1");
+            context.setAuthenticator(new BasicAuthenticator("realm") {
+                public boolean checkCredentials(String username, String password) {
+                    if (useHeader) {
+                        return username.equals("user") && password.equals("pwd");
+                    } else {
+                        return username.equals("serverUser") && password.equals("serverPwd");
+                    }
+                }
+            });
+            h2s.start();
+
+            int port = h2s.getAddress().getPort();
+            ServerAuth sa = new ServerAuth();
+            var plainCreds = "user:pwd";
+            var encoded = java.util.Base64.getEncoder().encodeToString(plainCreds.getBytes(US_ASCII));
+
+            URI uri = URIBuilder.newBuilder()
+                 .scheme("https")
+                 .host(addr.getHostAddress())
+                 .port(port)
+                 .path("/test1/foo.txt")
+                 .build();
+
+            HttpClient.Builder builder = HttpClient.newBuilder()
+                    .sslContext(ctx)
+                    .executor(ex);
+
+            builder.authenticator(sa);
+            client = builder.build();
+
+            HttpRequest req = HttpRequest.newBuilder(uri)
+                    .version(HttpClient.Version.HTTP_2)
+                    .header(useHeader ? "Authorization" : "X-Ignore", AUTH_PREFIX + encoded)
+                    .GET()
+                    .build();
+
+            HttpResponse<String> resp = client.send(req, HttpResponse.BodyHandlers.ofString());
+            if (useHeader) {
+                assertTrue(resp.statusCode() == 200, "Expected 200 response");
+                assertTrue(h.authValue().equals(encoded), "Expected user set header to be set");
+                assertTrue(!sa.wasCalled(), "Expected authenticator not to be called");
+                System.out.println("h2Test: using user set header OK");
+            } else {
+                assertTrue(resp.statusCode() == 200, "Expected 200 response");
+                assertTrue(!h.authValue().equals(encoded), "Expected user set header to not be set");
```

Comment:
It would be good to have a third scenario where the provided user header are bad (e.g. wrong password) and we verify that 401 is returned and that the request is not retried with the authenticator.